### DISCUSSION TITLE
Different cache logic for next-auth

### DISCUSSION THREAD

**Discussion by maiconcarraro** [serwist/serwist#28] [Q&A] [answered]
Hey, I recently migrated from next-pwa to serwist, so far so good, then I noticed some increase in the next-auth `OAuthCallback` **error**. 

I was doing some comparison and the cache files are different, so I wounder what was the reason to change from:

(return false for any /api/auth)
https://github.com/shadowwalker/next-pwa/blob/1e6af5fa9f6a807930f451adb5ab5078d5cda451/cache.js#L136

to

(return false only for /api/auth/callback)
https://github.com/serwist/serwist/blob/008d7945ba294854ea7f4cd00d393edb6a5925bc/packages/next/src/index.browser.ts#L144

For now I'm passing the previous logic and it seems stable (still monitoring), but posting here to get + context.

**Comment by DuCanhGH** [comment]
To be honest, even I am fairly confused about this (I, frankly, have never used `next-auth`). If you look at the comment right above the if branch, you can see that it tells us to exclude /api/auth/callback/\*, yet the if branch used to exclude /api/auth/\*, same as `next-pwa`. This was changed to match the comment (in [a `@ducanh2912/next-pwa` commit](https://github.com/DuCanhGH/next-pwa/commit/974326eb21b31b550b0d5b663cc35a2ebc19344c) rather than a Serwist one - not by me, mind you).

Not sure what it should be, but I will trace the change to /api/auth/\*, investigate the problem with `next-auth`, and change certain things should I see fit. The week after next week, perhaps. I'm kinda busy with other stuff at the moment...

**Reply by maiconcarraro** [reply]
I'm doing some investigation yet, will share updates as well

**Reply by maiconcarraro** [reply]
@DuCanhGH while I was doing the investigation some requests were sent w/ `Service-Worker-Navigation-Preload` header, I know some providers can complain depending on which headers are allowed or not, but the problem still happen even after disabling serwist... so not related (could be the reason for some situations, but definitely not all)

at the point I'm right now, I noticed it's doing double requests, so its something either on next-auth or next router causing this trigger, which is not related to this repo... should I close this discussion? (I can keep the discussion because of the original question still)

**Reply by DuCanhGH** [reply]
@maiconcarraro you do you :) I've not gotten around to investigating, so I'm quite in the dark about that.

**Reply by maiconcarraro** [reply]
@DuCanhGH is there a way to exclude specific routes for `navigationPreload`? can't find in docs

**Reply by maiconcarraro** [reply]
I have the change to disable cache for `api/auth` routes, but it didnt fix... so I was checking why it had a callback making twice (which invalidates the authorization code on 2nd attempt) and everytime the first request has `Service-Worker-Navigation-Preload`, I did some changes in my codebase to check if there are incoming `Service-Worker-Navigation-Preload` requests for api/auth and returning 403 error before executing any code... it solved the situation ✅ 

I could disable the navigationPreload, but if I can disable specific routes for navigation preload would be better

**Reply by DuCanhGH** [reply]
@maiconcarraro just add a new `runtimeCaching` entry that matches your routes and disable it there.

**Reply by maiconcarraro** [reply]
I had it, is there a way to debug to make sure the current service worker is updated w/ the cache config? I have return false for any `/api/auth/*` 🤔

**Reply by DuCanhGH** [reply]
@maiconcarraro just unregister the service worker and reload 💀

**Reply by maiconcarraro** [reply]
@DuCanhGH maybe the issue I had above w/ runtimeCaching was related to your new fix? https://github.com/serwist/serwist/commit/6a1986cff4d004d240c49b2fbc5775e38d29ee25 I'm going to check out

**Reply by DuCanhGH** [reply]
@maiconcarraro just asking, but why does caching "/api/auth/\*" cause issues? Looking at your described solution, I guess that it is due to navigation preloading. Unfortunately, I'm not aware of a way to opt out of that feature for specific routes at the moment. This is because `@serwist/strategies.StrategyHandler.fetch` unconditionally handles `event.preloadResponse` if it is available. I believe that your only way out is to disable navigation preloading altogether. Not sure why that fixes the issue though, since that feature is meant for GET navigations... Is this related to pages like "/api/auth/signin/:provider", which probably fetch a CSRF token from "/api/auth/csrf" on load?

Sorry for having not responded for such a long time. I've been snow under for the past few months, so it is fairly difficult to keep track of things. Not that I couldn't even write a reply, it is just that I forgot about this issue altogether...

**Reply by kdurek** [reply]
By the way, `lucia-auth` similarly didn't work, what fixed it, was comment below

**Comment by tony19** [comment, marked as answer]
I noticed the same thing. To disable caching for all `/api/auth/` routes, you have to add a `NetworkOnly` rule in the `runtimeCaching` array **before** the `defaultCache`. Otherwise, the `defaultCache` contains a `NetworkFirst` entry for `/api/` which would take precedence over your entry.

This worked for me:

```typescript
// src/app/sw.ts
import { defaultCache } from '@serwist/next/browser'
import { installSerwist } from '@serwist/sw'

installSerwist({
  //...
  runtimeCaching: [
    {
      urlPattern: /\/api\/auth\/.*/,
      handler: 'NetworkOnly',
      options: {}, // needed to avoid serwist crash
    },
    ...defaultCache,
  ],
})

```

**Reply by kdurek** [reply]
This fixed my issue on serwist 9.0.0

```ts
// src/app/sw.ts
import { ExpirationPlugin } from '@serwist/expiration';
import { defaultCache } from '@serwist/next/browser'
import { NetworkOnly } from '@serwist/strategies';
import { installSerwist } from '@serwist/sw'

installSerwist({
  //...
  runtimeCaching: [
    {
      matcher: /\/api\/auth\/.*/,
      handler: new NetworkOnly({
        plugins: [
          new ExpirationPlugin({
            maxEntries: 16,
            maxAgeSeconds: 24 * 60 * 60, // 24 hours
          }),
        ],
        networkTimeoutSeconds: 10, // fallback to cache if API does not response within 10 seconds
      }),
    },
    ...defaultCache,
  ],
})
```
